### PR TITLE
RSpec

### PR DESCRIPTION
**Pull request by TiagoPR**
Why:

* Test functionalities more easily and that everything is working. Possibly add it to the workflow.

This change addresses the need by:

* Adding the RSpec gem and setting it up. Creating some tests for the controllers that we have. Having the bin/rspec binstub.

### CODE REVIEW

File: spec/controllers/authentication_controller_spec.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+require "rails_helper"
+
+RSpec.describe AuthenticationController, type: :controller do
```
Comment (zamith) [inline spec/controllers/authentication_controller_spec.rb:3]: Controller tests are not great and we usually don't use them, because they skip the router and all middleware. Instead, I think it's best if you look at [requests specs](https://medium.com/@lcriswell/rails-api-request-specs-with-rspec-effeac468c4e).

PS: For now, you are forbidden from using `before`, `let` or `let!`. Later on we can explore why they are usually bad

File: spec/controllers/questions/answers_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
+require "rails_helper"
+
+RSpec.describe Questions::AnswersController, type: :controller do
+  describe "GET index" do
+    it "returns a boolean" do
+      user = User.create(email: "email@gmail.com", password: "password")
```
Comment (zamith) [inline spec/controllers/questions/answers_controller_spec.rb:6]: Take a look at factories with [factory_bot](https://github.com/thoughtbot/factory_bot)

Comment (zamith) [reply spec/controllers/questions/answers_controller_spec.rb:6]: Look at [4 phase tests](https://thoughtbot.com/blog/four-phase-test) and try to have an empty line between each phase in your tests.

**Review by zamith — CHANGES_REQUESTED** [review]

File: spec/controllers/authentication_controller_spec.rb (hunk 1 above)
Comment (TiagoPR) [reply spec/controllers/authentication_controller_spec.rb:3]: Now I'm curious why not to use `before`, it seems like it helps not to have to copy code again to test different things, like right now I have two different tests in which I had to do the get two times

Comment (zamith) [reply spec/controllers/authentication_controller_spec.rb]: Yes, and that's exactly the type of use cases where you don't want to use `before`. The `get` is an integral part of the test, meaning you can't understand the test without it. If you move it out, it means that in order to read the test you have to go up and down the file. Now image the file has over 500 lines (which is common), it's really hard to understand a test.

Tests should be optimized for being easy to understand when looking at a single one. If you ever heard the term DRY, for tests you want the opposite. It's ok to repeat yourself if it makes the test clearer. 

If you use `before` and `let` inside of different `context` or `describe` blocks it becomes even messier to read.

To sum up, it looks cool for a blog post when your tests are all one liners, but it makes almost impossible to work with, in my experience.

**Review by zamith — CHANGES_REQUESTED** [review]
Still not using request specs, right?

**Comment by TiagoPR** [comment]
> Still not using request specs, right?

I followed the link you sent, didn't fully understand the difference because it still used the controller 🧐

**Comment by TiagoPR** [comment]
@zamith Ok just saw another article about it, [https://medium.com/just-tech/rspec-controller-or-request-specs-d93ef563ef11](Controller vs Request Specs), going to change it!

File: spec/controllers/questions/answers_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
+require "rails_helper"
+
+RSpec.describe "Answers of a question", type: :request do
+  describe "GET index" do
+    it "returns http success" do
+      user = create(:user, email: "example2@gmail.com")
+      question = create(:question)
+      create(:answer, user: user, question: question)
+      post "/auth/login", params: {email: user.email, password: user.password}
```
Comment (zamith) [inline spec/controllers/questions/answers_controller_spec.rb:9]: This is something you are going to need a lot, so it's best to extract into a helper that returns the token, for example.

File: spec/controllers/week_question_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,20 @@
+require "rails_helper"
+
+RSpec.describe "Weekly Question", type: :request do
+  describe "GET show" do
+    it "returns a 200" do
```
Comment (zamith) [inline spec/controllers/week_question_controller_spec.rb:5]: Here you can not only change the verb, but also make it more descriptive.

```suggestion
    it "provides the weekly question for the given user" do
```

File: spec/factories/users.rb [hunk 5]
```diff
@@ -0,0 +1,6 @@
+FactoryBot.define do
+  factory :user do
+    email { "example@gmail.com" }
```
Comment (zamith) [inline spec/factories/users.rb:3]: This should use a [sequence](https://github.com/thoughtbot/factory_bot/blob/main/GETTING_STARTED.md#sequences), no?

**Review by zamith — CHANGES_REQUESTED** [review]

File: spec/controllers/questions/answers_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,19 @@
+require "rails_helper"
+
+RSpec.describe "Answers of a question", type: :request do
+  describe "GET index" do
+    it "returns http success" do
+      user = create(:user, email: "example2@gmail.com")
```
[the remark this answers is not in the document]
Comment (TiagoPR) [reply spec/controllers/questions/answers_controller_spec.rb:6]: I made it specifically because it complained of having another user with that email already because of the unique field. So I used it this way but should I define another factory of another user?

[the remark this answers is not in the document]
Comment (TiagoPR) [reply spec/controllers/questions/answers_controller_spec.rb:6]: It gave me an error because of email already taken but show saw that was a line I had before of create(:user) and created again another user, yea just forgot to remove and change.

File: spec/factories/users.rb (hunk 5 above)
Comment (TiagoPR) [reply spec/factories/users.rb:3]: Oh ok! That's perfect, didn't know about it! Thanks

File: spec/controllers/authentication_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,22 @@
+require "rails_helper"
+
+RSpec.describe "Authentication", type: :request do
```
[the remark this answers is not in the document]
Comment (TiagoPR) [reply spec/controllers/authentication_controller_spec.rb:3]: I think that if I remove the type if will think to be a controller spec, I think I have my files on the correct path on the `spec/controllers`

File: spec/controllers/authentication_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+  describe "GET login" do
+    it "returns a 200" do
+      user = create(:user)
+
+      post "/auth/login", params: {email: user.email, password: user.password}
```
[the remark this answers is not in the document]
Comment (TiagoPR) [reply spec/controllers/authentication_controller_spec.rb:8]: I've been searching and I'm not really sure how to do it, could you provide an example or a link with more information?

File: spec/controllers/questions/answers_controller_spec.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (zamith) [reply spec/controllers/questions/answers_controller_spec.rb]: yeah, because of what I said below about sequences.

File: spec/controllers/authentication_controller_spec.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (zamith) [reply spec/controllers/authentication_controller_spec.rb:3]: it should be on `spec/requests` I think

[the remark this answers is not in the document]
Comment (TiagoPR) [reply spec/controllers/authentication_controller_spec.rb:3]: Yup, it was that

File: spec/controllers/authentication_controller_spec.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (TiagoPR) [reply spec/controllers/authentication_controller_spec.rb]: Think I got it

File: config/routes.rb [hunk 9]
```diff
@@ -8,7 +8,7 @@
   end
   resources :answers
   resources :weekly_friends, only: [:show], controller: "weekly_friends"
-  post "/auth/login", to: "authentication#login"
-  get "/weekly_question", to: "week_question#show"
+  post "/auth/login", to: "authentication#login", as: "login"
```
Comment (zamith) [inline config/routes.rb:11]:
```suggestion
  post "/auth/login", to: "authentication#login", as: :login
```
It's common to use atoms for things that are identifiers, such as hash keys and route names

File: spec/requests/week_question_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,18 @@
+require "rails_helper"
+
+RSpec.describe "Weekly Question" do
+  context "GET show" do
+    it "provides the weekly question for the given user" do
+      user = create(:user)
+      question = create(:question)
+      weekly_question = WeeklyQuestion.create(question: question, week: Date.today.end_of_week + 1.day)
+      token = sign_in_as(user)
+      header = {"Authorization" => "Bearer #{token}"}
```
Comment (zamith) [inline spec/requests/week_question_controller_spec.rb:10]: You might consider have the helper return the header, or have another helper that takes a token and returns the headers. If we want to change the header for some reason we'd have to change it in many tests, otherwise.

**Review by zamith — APPROVED** [review]

**Comment by zamith** [comment]
Don't forget to run the tests on the PRs and main now
